### PR TITLE
Add functions to use symmetric encryption with provided IVs [MAID-2918]

### PR DESCRIPTION
**Pull request by nbaksalyar**
* Add functions to use symmetric encryption with custom initialisation vectors. We use a randomly generated unique nonce as an IV by default.
* Also add a function to create a symmetric encryption key out of a byte array.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -71,8 +71,18 @@ use serde::{de::DeserializeOwned, Serialize};
...
+pub const SYMMETRIC_KEY_SIZE: usize = secretbox::KEYBYTES;
+/// Size of a public signing key.
+pub const PUBLIC_SIGN_KEY_SIZE: usize = 32;
+
 /// Represents public signature key.
-pub type PublicSignKey = [u8; 32];
+pub type PublicSignKey = [u8; PUBLIC_SIGN_KEY_SIZE];
+
+/// Initialisation vector.
+pub type Iv = [u8; IV_SIZE];
```
Comment (povilasb) [inline src/lib.rs]: Maybe have a longer name for a public type? e.g. `InitVec`, `InitializationVec`, etc.

File: src/lib.rs [hunk 2]
```diff
@@ -352,6 +362,14 @@ impl SymmetricKey {
         }
     }
 
+    /// Create a new key from bytes.
```
Comment (povilasb) [inline src/lib.rs:377]: Just curious, in what cases do we store our encryption key serialized? :)

Comment (nbaksalyar) [reply src/lib.rs:377]: @povilasb, the access container in the Authenticator actually stores symmetric encryption keys in the serialised form (to be more exact, it stores the [`MDataInfo`](https://github.com/maidsafe/safe_client_libs/blob/4f8b40f53fad1ff9b874085dd10c51ad0090c42a/safe_core/src/client/mdata_info.rs) structs that can point to private containers, along with their encryption key).

File: src/lib.rs [hunk 3]
```diff
@@ -364,6 +382,17 @@ impl SymmetricKey {
         self.encrypt_bytes(&serialise(plaintext)?)
     }
 
+    /// Encrypts bytestring `plaintext` using initialisation vector `iv`.
+    ///
+    /// Returns ciphertext in case of success.
+    pub fn encrypt_bytes_iv(&self, plaintext: &[u8], iv: Iv) -> Result<Vec<u8>, Error> {
```
Comment (povilasb) [inline src/lib.rs]: Even though we provide custom nonce, does the requirement of using it only once still apply? If so, I think we should explicitly state that to make it harder to misuse this function.

Comment (nbaksalyar) [reply src/lib.rs]: Great catch, thanks!

However, in future versions we'll replace the symmetric encryption algo with Miscreant (which is [_nonce-misuse resistant_](https://github.com/miscreant/miscreant/wiki/Nonce-Reuse-Misuse-Resistance)) and this won't apply in that case.

File: src/lib.rs [hunk 4]
```diff
@@ -364,6 +382,17 @@ impl SymmetricKey {
...
+        Ok(secretbox::seal(
```
Comment (povilasb) [inline src/lib.rs]: If we can never fail, why return `Error`?

File: src/lib.rs (hunk 3 above)
Comment (povilasb) [reply src/lib.rs]: Ciphers with nonce reuse misuse resistance, such as the AES-SIV and AES-PMAC-SIV ciphers provided by this library, do not fail catastrophically in this case: the only information these ciphers leak is that if the same message is encrypted with the same nonce twice under the same key, we obtain the same ciphertext. **Though this information is still potentially useful to attackers and thus we should still avoid reusing nonces if possible,**

Seems like we still should discourage `safe_crypto` users from reusing nonces.

File: src/lib.rs [hunk 5]
```diff
@@ -92,6 +102,11 @@ pub fn hash(data: &[u8]) -> [u8; 32] {
     hashing_impl::sha3_256(data)
 }
 
+/// Generate a new unique nonce.
+pub fn gen_nonce() -> Nonce {
```
Comment (ustulation) [inline src/lib.rs:106]: what's the use of this - is it not misleading when you already have things like [these](https://github.com/maidsafe/safe_crypto/pull/15/files#diff-b4aea3e418ccdb71239b96952d9cddb6L288) ?

**Review by ustulation — CHANGES_REQUESTED** [review]

**Pull request closed without merging**